### PR TITLE
Add implementation of maximize.

### PR DESCRIPTION
**Pull request by toshihikoyanase**
This PR adds logic to support maximization. With this PR, users can write maximization simply as follows:

```python
import sklearn.datasets
import sklearn.model_selection
import sklearn.svm

def objective(trial):
    iris = sklearn.datasets.load_iris()
    x, y = iris.data, iris.target

    svc_c = trial.suggest_loguniform('svc_c', 1e-10, 1e10)
    classifier_obj = sklearn.svm.SVC(C=svc_c)
    score = sklearn.model_selection.cross_val_score(classifier_obj, x, y, n_jobs=-1, cv=3)
    accuracy = score.mean()
    return accuracy  # Directly use accuracy as an objective value.


if __name__ == '__main__':
    import optuna
    study = optuna.create_study(direction='maximize')  # Set study direction to maximize.
    study.optimize(objective, n_trials=100)
    print(study.best_trial)
```

I have two concerns.

1) This implementation assumes that users use the same direction both for samplers and pruners. This assumption makes implementation simple, but it will suppress flexibility. For instance, [this example](https://github.com/pfnet/optuna/blob/master/examples/pruning/tensorflow_estimator_integration.py) in current master uses validation accuracy for the pruner and validation error for the sampler. On the other hand, validation accuracy is used for both the pruner and the sampler in this PR.

2) Also, this PR contains an API change.  `is_higher_better` argument of TensorFlowPruningHook.__init__() was disabled [here](https://github.com/pfnet/optuna/pull/331/files#diff-6777d6b27690da66a18f118b10e4a1dcR57) to ensure that both samplers and pruners use the same study direction.

### CODE REVIEW

File: optuna/integration/lightgbm.py [hunk 1]
```diff
@@ -72,11 +72,9 @@ def __call__(self, env):
             if valid_name != target_valid_name or metric != self.metric:
                 continue
 
-            # TODO(ohta): Deal with maximize direction
             if is_higher_better:
-                raise ValueError(
-                    'Pruning using metrics to be maximized has not been supported yet '
-                    '(validation_name: {}, metric: {}).'.format(valid_name, metric))
+                assert self.trial.storage.get_study_direction(self.trial.study_id) == \
```
Comment (sile) [inline optuna/integration/lightgbm.py]: I think an `else` block like the following is needed for checking the consistency when `is_higher_better == False`:
```python
else:
    assert self.trial.storage.get_study_direction(self.trial.study_id) == optuna.structs.StudyDirection.MINIMIZE
```
```

Comment (toshihikoyanase) [reply optuna/integration/lightgbm.py]: You're right. I fixed it.
Thank you for pointing it out.

Comment (g-votte) [reply optuna/integration/lightgbm.py]: Since this is user facing when the assertion fails, we might want to clarify the reason with an error message.

```python
else:
    if self.trial.storage.get_study_direction(self.trial.study_id) != optuna.structs.StudyDirection.MINIMIZE:
        raise ValueError(...)
```

**Comment by g-votte** [comment]
Please fix the document of `Study` that mentions `maximize` is not supported.
https://github.com/pfnet/optuna/pull/331/files#diff-4e020643693eb296936f5dd7194bd6d6L481

File: optuna/integration/lightgbm.py (hunk 1 above)
Comment (toshihikoyanase) [reply optuna/integration/lightgbm.py]: I agree with you. I added the following error message. Thank you. 

> 'The intermediate values are inconsistent with the objective values in terms of study directions. Please specify a metric to be minimized for LightGBMPruningCallback.'

**Comment by toshihikoyanase** [comment]
> https://github.com/pfnet/optuna/pull/331/files#diff-4e020643693eb296936f5dd7194bd6d6L481

I remove the sentence. Thank you for your careful review.

**Comment by sile** [comment]
I found a TODO comment that seems related to the maximization support: [test_storage.py#L131](https://github.com/pfnet/optuna/blob/add-maximize/tests/storages_tests/rdb_tests/test_storage.py#L131)

**Comment by toshihikoyanase** [comment]
> I found a TODO comment that seems related to the maximization support: [test_storage.py#L131](https://github.com/pfnet/optuna/blob/add-maximize/tests/storages_tests/rdb_tests/test_storage.py#L131)

Thank you for letting me know.
I added an additional trial to the study and removed the TODO comment.

File: optuna/integration/tensorflow.py [hunk 2]
```diff
@@ -43,25 +46,28 @@ class TensorFlowPruningHook(SessionRunHook):
...
         self.trial = trial
         self.estimator = estimator
         self.current_summary_step = -1
         self.metric = metric
-        self.is_higher_better = is_higher_better
         self.global_step_tensor = None
         self.timer = tf.train.SecondOrStepTimer(every_secs=None, every_steps=run_every_steps)
 
+        if is_higher_better is not None:
+            raise ValueError('Please do not use is_higher_better argument of'
```
Comment (g-votte) [inline optuna/integration/tensorflow.py]: We might want to clarify the reason, instead of just saying "please do not use...". E.g., `is_higher_better argument is obsolete since Optuna 0.9.0.`

Comment (toshihikoyanase) [reply optuna/integration/tensorflow.py]: Thank you for your suggestion. I added the sentence you proposed.
